Allow specify dynamic templates in bulk request

dnhatn opened the pull request:
This change allows users to specify dynamic templates in a bulk request.

```js
PUT myindex
{
  "mappings": {
    "dynamic_templates": [{
      "time_histograms": {
        "mapping": {
          "type": "histogram",
          "meta": {
            "unit": "s"
          }
        }
      }
    }]
  }
}
```

```js
// Tells ES to use the dynamic template named `time_histograms` for field `response_times`
// in the index request
POST myindex/_bulk
{ "index": { "dynamic_templates": { "response_times": "time_histograms" } } }
{ "@timestamp": "2020-08-12", "response_times": { "values": [1, 10], "counts": [5, 1] }}
```

Closes #61939

dnhatn committed:
Add dynamic mapping type hint to bulk request

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

javanna reviewed (commented):
I left some questions. The approach looks good code-wise overall, my comments are mostly about the API.

Another one: does the matched "type" become the content of the dynamic_type variable? Or should it even? I am wondering myself if this should be considered and promoted as a type. To me it is more of an arbitrary name that defined dynamic templates can match against. Should we rather call it `match_mapping_hint`?

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
Remove TODO

dnhatn commented on server/src/main/java/org/elasticsearch/index/termvectors/TermVectorsService.java:
```diff
@@ -323,8 +323,9 @@ private static ParsedDocument parseDocument(IndexShard indexShard, String index,
                                                 XContentType xContentType, String routing) {
         MapperService mapperService = indexShard.mapperService();
         DocumentMapperForType docMapper = mapperService.documentMapperWithAutoCreate();
+        // TODO:?
```
Thanks, fixed in https://github.com/elastic/elasticsearch/pull/69948/commits/27a538bfb2f679a2e3d9f078f5f3e724fd66b69d

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
match_mapping_hint

dnhatn committed:
fix doc

dnhatn committed:
fix test

dnhatn commented:
@javanna Thank you for the review. I tweaked the API using "match_mapping_hint" and I found it's simpler and less confusing. I have updated the PR. Can you take another look?

axw commented:
@dnhatn apologies if this is obvious: will this be immediately usable through an Ingest script processor (and if so how?), or would that come later?

dnhatn commented:
> can users use one of these types as hint in their bulk requests, in which case they won't have to do define anything on the dynamic template side of things? Or should hints always be arbitrary and explicitly defined in the dynamic template?

That's a good question. Let's discuss it after this iteration.

jimczi reviewed (approved):
The change looks great @dnhatn ! I left some comments but I think the feature is ready.

javanna reviewed (changes requested):
I left some more comments, nothing major though. I like how simple the implementation, thanks for iterating on it. Would it make sense to add docs as part of this PR or do you plan to do it as a followup?

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
dynamicMappingHints -> dynamicMatchMappingHints

dnhatn committed:
javadocs

dnhatn committed:
stronger yaml test

dnhatn committed:
match mapping hint -> mapping hint

dnhatn committed:
add index request test

dnhatn committed:
revert tests

dnhatn committed:
Remove left over

dnhatn committed:
use existing methods for test

dnhatn committed:
add test for bulk request

dnhatn commented on rest-api-spec/src/main/resources/rest-api-spec/test/bulk/11_type_hints.yml:
```diff
@@ -0,0 +1,46 @@
+          - index:
+              _index: test_index
+              _id: id_1
+              match_mapping_hints:
```
I think we should, but we target bulk requests first in this PR.

dnhatn commented on rest-api-spec/src/main/resources/rest-api-spec/test/bulk/11_type_hints.yml:
```diff
@@ -0,0 +1,46 @@
+          - index:
+              _index: test_index
+              _id: id_2
+              match_mapping_hints:
```
+1, I renamed it in https://github.com/elastic/elasticsearch/pull/69948/commits/12a87ed90194f033cee6c125847fd3455d940cc4.

dnhatn commented on server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -146,6 +148,11 @@ public IndexRequest(@Nullable ShardId shardId, StreamInput in) throws IOExceptio
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+            dynamicMatchMappingHints = in.readMap(StreamInput::readString, StreamInput::readString);
+        } else {
+            dynamicMatchMappingHints = Map.of();
```
fixed in https://github.com/elastic/elasticsearch/pull/69948/commits/f99abcdd49000f735ff258d5a4fb93ee8a705bd3

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:
```diff
@@ -668,4 +668,26 @@ public Object externalValue() {
+     * @return the matching template; otherwise returns null
+     * @throws MapperParsingException if the given field has a type hint, but no template matches
+     */
+    public final DynamicTemplate findDynamicTemplate(String name, DynamicTemplate.XContentFieldType matchType) {
```
Good catch, I removed the leftover in https://github.com/elastic/elasticsearch/pull/69948/commits/f2e6c0f3bd89d475dd58c760fc2cf9a7a0ddfde1.

dnhatn committed:
more rename

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/SourceToParse.java:
```diff
@@ -57,6 +62,10 @@ public String id() {
         return this.routing;
     }
 
+    public Map<String, String> dynamicMappingHints() {
```
++, adjusted

dnhatn commented on server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplateTests.java:
```diff
@@ -116,8 +118,7 @@ public void testParseUnknownMatchType() {
                 () -> DynamicTemplate.parse("my_template", templateDef2));
-        assertEquals("No field type matched on [text], possible values are [object, string, long, double, boolean, date, binary]",
-                e.getMessage());
+        assertThat(e.getMessage(), containsString("No field type matched on [text], possible values are ["));
```
reverted in https://github.com/elastic/elasticsearch/pull/69948/commits/d126606a15bdef9a26516d1a93b72cf6ffce5634

dnhatn commented on server/src/internalClusterTest/java/org/elasticsearch/index/mapper/DynamicMappingIT.java:
```diff
@@ -77,6 +95,14 @@ private static void assertMappingsHaveField(GetMappingsResponse mappings, String
         assertTrue("Could not find [" + field + "] in " + typeMappingsMap.toString(), properties.containsKey(field));
     }
 
+    private static void assertFieldMappingType(String index, String field, String expectedType) {
```
Yes, I removed it.

dnhatn commented on server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:
```diff
@@ -52,6 +52,7 @@
     private static final ParseField IF_SEQ_NO = new ParseField("if_seq_no");
     private static final ParseField IF_PRIMARY_TERM = new ParseField("if_primary_term");
     private static final ParseField REQUIRE_ALIAS = new ParseField(DocWriteRequest.REQUIRE_ALIAS);
+    private static final ParseField MATCH_MAPPING_HINTS = new ParseField("match_mapping_hints");
```
++, added in https://github.com/elastic/elasticsearch/pull/69948/commits/5679e5edb70d97bf0b3cedd2e2ffe703ff17cbf5

dnhatn commented on server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -655,6 +662,13 @@ private void writeBody(StreamOutput out) throws IOException {
+            out.writeMap(dynamicMatchMappingHints, StreamOutput::writeString, StreamOutput::writeString);
+        } else {
+            if (dynamicMatchMappingHints.isEmpty() == false) {
+                throw new IllegalStateException("Dynamic mapping type hints requires all nodes in the cluster on 8.0 or later");
```
Added in https://github.com/elastic/elasticsearch/pull/69948/commits/f99abcdd49000f735ff258d5a4fb93ee8a705bd3.

dnhatn commented on server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java:
```diff
@@ -1037,6 +1043,101 @@ public void testDynamicDottedFieldNameLongArrayWithParentTemplate() throws Excep
+        assertThat(getFieldTypes(doc, field), arrayContainingInAnyOrder(LatLonPoint.TYPE, LatLonDocValuesField.TYPE));
+
+        doc = mapper.parse(source("1", b -> b.field(field, new String[]{"41.12,-71.34", "43,-72.34"}), null, Map.of(field, "geo_point")));
+        assertThat(getFieldTypes(doc, field), arrayContainingInAnyOrder(
```
I updated to use the existing method instead in https://github.com/elastic/elasticsearch/pull/69948/commits/00c6454fbdce60ad8a3e02ad24ce156b10502da6.

dnhatn commented on server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java:
```diff
@@ -1037,6 +1043,101 @@ public void testDynamicDottedFieldNameLongArrayWithParentTemplate() throws Excep
+        assertThat(getFieldTypes(doc, field), arrayWithSize(2));
+        MapperParsingException error = expectThrows(MapperParsingException.class, () ->
+            mapper.parse(source("1", b -> b.field(field, "hello"), null, Map.of(field, "new_type"))));
+        assertThat(error.getMessage(), containsString("Can't find template for matching hint [new_type] of field"));
```
I adjusted the test https://github.com/elastic/elasticsearch/pull/69948/commits/00c6454fbdce60ad8a3e02ad24ce156b10502da6.

dnhatn commented on server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java:
```diff
@@ -1037,6 +1043,101 @@ public void testDynamicDottedFieldNameLongArrayWithParentTemplate() throws Excep
+                {
+                    b.startObject("keywords");
+                    {
+                        b.field("match_mapping_hint", "keyword");
```
I think we should :)

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/SourceToParse.java:
```diff
@@ -57,6 +62,10 @@ public String id() {
         return this.routing;
     }
 
```
added in https://github.com/elastic/elasticsearch/pull/69948/commits/de97281d5071df6350dcdc35e6713b5445698ba4

dnhatn commented on rest-api-spec/src/main/resources/rest-api-spec/test/bulk/11_type_hints.yml:
```diff
@@ -0,0 +1,46 @@
+                bottom_right:
+                  lat: 40
+                  lon: -74
+  - match: { hits.total.value: 2 }
```
+1, I added that test in https://github.com/elastic/elasticsearch/pull/69948/commits/635c3957c1b4f759f9b2bd19bcfdbfbaea747a70

dnhatn commented on server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -712,4 +726,13 @@ public IndexRequest setRequireAlias(boolean requireAlias) {
+        this.dynamicMatchMappingHints = Objects.requireNonNull(dynamicMatchMappingHints);
+        return this;
+    }
+
```
added in https://github.com/elastic/elasticsearch/pull/69948/commits/de97281d5071df6350dcdc35e6713b5445698ba4

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:
```diff
@@ -668,4 +668,26 @@ public Object externalValue() {
+     */
+    public final DynamicTemplate findDynamicTemplate(String name, DynamicTemplate.XContentFieldType matchType) {
+        final String pathAsString = path().pathAsText(name);
+        final String matchHint = sourceToParse().dynamicMappingHints().get(pathAsString);
```
+1, I will do it in a follow-up.

dnhatn commented:
> Would it make sense to add docs as part of this PR or do you plan to do it as a followup?

Yes, I will add docs in a follow-up. 

@javanna Thank you for you review. Can you take another look?

dnhatn committed:
stylecheck

javanna commented on rest-api-spec/src/main/resources/rest-api-spec/test/bulk/11_mapping_hints.yml:
```diff
@@ -0,0 +1,122 @@
+  - match: { items.0.index.error.reason: "Can't find template for mapping hint [bar_foo] of field [new_location]"}
+  - match: { items.1.index.status: 400 }
+  - match: { items.1.index.error.type: mapper_parsing_exception }
+  - match: { items.1.index.error.reason: "Can't find template for mapping hint [foo_bar] of field [new_location]"}
```
nit: shall we specify "dynamic template". I am trying to see the error from a user perspective and make sure it is clear.

javanna commented on server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:
```diff
@@ -251,12 +261,18 @@ public void parse(
                             throw new IllegalArgumentException("Update requests do not support versioning. " +
                                     "Please use `if_seq_no` and `if_primary_term` instead");
                         }
+                        // TODO: support dynamic mapping hints for update requests
```
is the plan to work on this as a follow-up?

dnhatn replied:
Yes, I will work on a follow-up for this.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -711,7 +712,12 @@ private static void parseCopy(String field, ParseContext context) throws IOExcep
+                    final Mapper fieldMapper = dynamic.getDynamicFieldsBuilder().createDynamicObjectMapper(context, paths[i]);
+                    if (fieldMapper instanceof ObjectMapper == false) {
+                        throw new MapperParsingException("Field [" + context.path().pathAsText(paths[i]) + "] must be an object; " +
+                            "but it's configured as [" + fieldMapper.typeName() + "] in dynamic templates");
```
Question: can this happen in reality? If so do we have a test for it? I am wondering if this became relevant with the previous iteration but it may no longer be the case now that hints are separate from types. Maybe we should rather have an assertion in createDynamicObjectMapper and cast there so that its return type can be the right one? This is unrelated to your PR though :)

javanna replied:
I still wonder about this added conditional ;)

dnhatn replied:
I added a test for this in https://github.com/elastic/elasticsearch/pull/69948/commits/a6f99710eda8c035acad03789f0684047509e36f.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -225,9 +231,11 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
     private final XContentFieldType xcontentFieldType;
     private final Map<String, Object> mapping;
     private final boolean runtimeMapping;
+    private final String matchHint;
```
did you mean `matchMappingHint` ? :)

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -247,7 +256,7 @@ public String pathMatch() {
     }
 
-    public boolean match(String path, String name, XContentFieldType xcontentFieldType) {
+    public boolean match(String path, String name, String hint, XContentFieldType xcontentFieldType) {
```
call the argument mappingHint?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:
```diff
@@ -668,4 +668,26 @@ public Object externalValue() {
+     */
+    public final DynamicTemplate findDynamicTemplate(String name, DynamicTemplate.XContentFieldType matchType) {
+        final String pathAsString = path().pathAsText(name);
+        final String matchHint = sourceToParse().mappingHints().get(pathAsString);
```
mappingHint ?

javanna reviewed (approved):
left some nits, but this looks ready to me besides those small naming comments I left. LGTM! thanks @dnhatn !

jpountz commented:
Agreed with @javanna that this looks great. I'm happily surprised with the low number of new lines of codes that this PR adds!

One thing I wonder about the API is whether we actually need a new parameter in dynamic templates, or if we could reuse the dynamic template name. For instance the example from the issue description uses about the same name for the template name and for the mapping hint, which looks like a good practice to me. Should we go further and match the index request's `mapping_hints` on dynamic template names rather than on their `match_mapping_hint`? We could change the parameter name to something like `dynamic_template_names` or `dynamic_template_name_hints` to make the connection clearer if we thought this is a better API.

Another thing I wonder is what is the right behavior in case of conflict between these hints and existing options of dynamic templates like `match_mapping_type` and `path_match`. For instance, say you have a dynamic template configured with `match_mapping_type: object` and then create an index request that is configured to use this dynamic template on a string field. My understanding of the current code is that we would ignore _all_ dynamic templates in such a case, even if there is one that matches on `match_mapping_type:string`. This behavior goes a bit against the intuition I have for a `hint`, which is that things should behave as if there had been no hint provided if there isn't a match, ie. we should take the template that matches on `match_mapping_type:string` rather than ignoring all dynamic templates? Having precedence rules in docs before merging would help me better think through these cases.

javanna commented:
> One thing I wonder about the API is whether we actually need a new parameter in dynamic templates, or if we could reuse the dynamic template name

that is a great idea, I wonder how I did not see that myself , thanks for bringing it up :)  I think that removes one level of indirection which is not needed. I would then use something like `dynamic_template_hints`. I suppose that `template_hints` is too short and unclear...

> Another thing I wonder is what is the right behavior in case of conflict between these hints and existing options of dynamic templates like match_mapping_type and path_match

I agree. I just realized that I am not entirely clear on the current behaviour, so maybe more testing would help identifying what is more sensible from a user perspective.

dnhatn commented:
@jpountz Thanks for chiming in :).

I will update the PR to use the dynamic template name instead.

> My understanding of the current code is that we would ignore all dynamic templates in such a case, even if there is one that matches on match_mapping_type:string.

We will throw an exception in this case and fail the related item of a bulk request. I think that's a good behavior? WDYT?

jpountz commented:
> We will throw an exception in this case and fail the related item of a bulk request. I think that's a good behavior? WDYT?

Oh right, sorry for the noise! I wonder if there are cases when this will not be practical, e.g. if the data shipper doesn't have full control over the data that gets indexed, maybe @axw has thoughts on this.

axw commented:
>>> My understanding of the current code is that we would ignore all dynamic templates in such a case, even if there is one that matches on match_mapping_type:string.
>>
>> We will throw an exception in this case and fail the related item of a bulk request. I think that's a good behavior? WDYT?
>
> Oh right, sorry for the noise! I wonder if there are cases when this will not be practical, e.g. if the data shipper doesn't have full control over the data that gets indexed, maybe @axw has thoughts on this.

Not sure I follow the above. Are we saying that if a hint is specified that matches no dynamic template, then the document will be failed if it _does_ match some other dynamic template rule?

That would be fine for our use case, as we only have a fairly limited set of dynamic templates that shouldn't match any fields for which we'll use hints. Although we don't know the field names ahead of time, we still control when hints will be applied, and there are some restrictions on the field names. Namely, they shouldn't collide with any well-defined fields like those in ECS, which is where we have path_match rules in place. If a user specifies a metric/field name that collides, which would lead to a hint that conflicts with a path_match, then IMO it's reasonable for the document to be failed.

However, I don't find this behaviour intuitive from the word "hint".

dnhatn commented:
@axw Thank you for your feedback.


> However, I don't find this behaviour intuitive from the word "hint".

Are you suggesting "hints" should be optional?

axw commented:
> Are you suggesting "hints" should be optional?

No, I was only commenting on the intuitiveness of the naming. Behaviour-wise, I would prefer to have strict control over how something is dynamically mapped: if a "hint" (dynamic template name) is specified and no such dynamic template exists, then fail the document. Maybe it could be made optional? Fail vs. fall back to existing dynamic template matching rules?

javanna commented:
heya @dnhatn based on the conversation around "conflicts" and behaviour above, I am thinking that it would be beneficial to work on the docs as part of this PR and highlight these different scenarios. Possibly also expand testing to cover those same scenarios. Would that work for you?

dnhatn commented:
@javanna Sure, I will add docs and conflict tests in this PR.

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
feedback

dnhatn committed:
revert

dnhatn committed:
Use template name

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
fix docs

dnhatn commented:
@javanna @jpountz  

I've updated the PR to use the dynamic template name as we discussed. 
Two notes on the implementation:

- Dynamic templates without match conditions are now [allowed](https://github.com/elastic/elasticsearch/blob/58af624d32f6672d3762832bc95b6b0110e96a94/server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java#L393-L398). Without this, dynamic templates can't be used exclusively for the name hints. We need to add some conditions to prevent a template from matching anything except fields used with the name hints.

- When using the name hints, other match conditions of a template are [ignored](https://github.com/elastic/elasticsearch/blob/58af624d32f6672d3762832bc95b6b0110e96a94/server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java#L259-L262) because users explicitly "choose" the template.

Would you please take another look? Thank you!

javanna commented on docs/reference/mapping/dynamic/templates.asciidoc:
```diff
@@ -14,6 +14,10 @@ name
 
+* if a dynamic template does not define `match_mapping_type` nor `match` nor
+`path_match`, then it won't match any field except fields are used with 
+name hints in a <bulk, bulk request>.
```
I guess it is still possible to have a template with e.g. `match: "geo*"` and point to it by name from a hint?

dnhatn replied:
That's correct. I added more tests for this in https://github.com/elastic/elasticsearch/pull/69948/commits/315c8da260c5f9d8f464ad2f31674782af32844a.

javanna commented on docs/reference/mapping/dynamic/templates.asciidoc:
```diff
@@ -14,6 +14,10 @@ name
 dotted path to the field
 
+* if a dynamic template does not define `match_mapping_type` nor `match` nor
+`path_match`, then it won't match any field except fields are used with 
```
I would slightly rephrase it to "then it won't match any field, but it can be referred to by name in `dynamic_template_hints` as part of a bulk request"

dnhatn replied:
++, I reworded it in https://github.com/elastic/elasticsearch/pull/69948/commits/d5a92310d866a74b032976f00cd6782608f3528a

javanna commented on docs/reference/docs/bulk.asciidoc:
```diff
@@ -738,3 +742,38 @@ The API returns the following result.
+{ "field" : "value2", "home_location": "41.12,-71.34"}
+----
+
+The bulk request creates two new fields `work_location` and `home_location` with type `geo_point` because 
```
replace because of with "according to" ?

dnhatn replied:
reworded in https://github.com/elastic/elasticsearch/pull/69948/commits/d5a92310d866a74b032976f00cd6782608f3528a.

javanna commented on docs/reference/rest-api/common-parms.asciidoc:
```diff
@@ -631,6 +631,15 @@ If `true`, the destination must be an <<indices-aliases,index alias>>. Defaults
+(Optional, map)
+A map of hints from the full name of fields to the name of <<dynamic-templates, dynamic templates>.
+Defaults to an empty map. If a name hint matches a dynamic template, then that template will be applied
+regardless of other match predicates defined in the template. And if a field is defined in the mapping,
```
If a field is **already** defined

javanna commented on docs/reference/rest-api/common-parms.asciidoc:
```diff
@@ -631,6 +631,15 @@ If `true`, the destination must be an <<indices-aliases,index alias>>. Defaults
+Defaults to an empty map. If a name hint matches a dynamic template, then that template will be applied
+regardless of other match predicates defined in the template. And if a field is defined in the mapping,
+then its name hint won't be used.
+end::bulk-dynamic-template-hints[]
```
shall we describe what happens when you associate a field with a non existing template? Based on this, I wonder (sorry!) if we should rename this parameter. `hint` may mislead users into thinking that it is best effort and only applied when possible, which I believe does not correspond with the behaviour. maybe now that we use the template name we could remove the hints suffix and go with 'dynamic_template`? Thoughts on this?

dnhatn replied:
Good suggestion. I removed "hint" from this PR entirely in https://github.com/elastic/elasticsearch/pull/69948/commits/d2dfda1c6ab2caf5c1e8cdb5c38ec78105fe6489.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -181,21 +181,24 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
         }
-
-        if (match == null && pathMatch == null && matchMappingType == null) {
-            throw new MapperParsingException("template must have match, path_match or match_mapping_type set " + conf.toString());
```
relaxing this makes sense to me. I wonder though about the behavior in a mixed cluster, where only new nodes can support the newer format, and older nodes would not accept it. I assume things will fail, but where and how?

dnhatn replied:
We have the validation here https://github.com/elastic/elasticsearch/blob/00b92c15d6e66f8e6fe33953a03edf9107c31c3a/server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java#L393-L398.

dnhatn replied:
It's the [smallest](https://github.com/elastic/elasticsearch/blob/321f46e1877d40d987294f28362619a322d92af3/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java#L673) version of non-client nodes.

dnhatn replied:
@javanna I pushed https://github.com/elastic/elasticsearch/pull/69948/commits/aef927e301f68b67ea3f44a3a1cbaee866377b17 to use the min node version of the current state instead.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -181,21 +181,24 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
+                xContentFieldTypes = XContentFieldType.values();
+            }
+        } else if (matchMappingType.equals("*")) {
+            xContentFieldTypes = XContentFieldType.values();
```
I wonder with the new logic, if it is counter intuitive that we add all types, knowing that runtime mappings are only applied to a subset of types. Maybe we should filter the types here in that case to only the ones that support runtime mappings? Then we shoud be able to remove the last condition from the match method as well as some logic from RootObjectMapper#validateDynamicTemplate? Do you think it makes sense?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -181,21 +181,24 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
-        if (matchMappingType != null && matchMappingType.equals("*") == false) {
-            xcontentFieldType = XContentFieldType.fromString(matchMappingType);
-            if (runtime && xcontentFieldType.supportsRuntimeField() == false) {
+        XContentFieldType[] xContentFieldTypes = new XContentFieldType[0];
```
nit: shall we make this final and initialize it in each branch instead?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -362,9 +374,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
-        } else if (match == null && pathMatch == null) {
+        if (xContentFieldTypes.length == 1) {
+            builder.field("match_mapping_type", xContentFieldTypes[0]);
+        } else if (xContentFieldTypes.length > 1 && match == null && pathMatch == null) {
```
maybe this deserves a comment: empty array means the user did not specify it, 1 means a specific type was specified, > 1 means we expanded it to the corresponding types (depending on whether mapping or runtime is specified) while the user provided `*`

dnhatn replied:
I added a comment

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:
```diff
@@ -668,4 +668,27 @@ public Object externalValue() {
+        }
+        if (matchTemplateName != null) {
+            throw new MapperParsingException(
+                "Can't find dynamic template for dynamic template name hint [" + matchTemplateName + "] of field [" + pathAsString + "]");
```
this is what I think makes it not a hint, that I was referring to above in another comment

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:
```diff
@@ -400,18 +390,19 @@ private static void validateDynamicTemplate(Mapper.TypeParser.ParserContext pars
+        if (parserContext.indexVersionCreated().before(Version.V_8_0_0)) {
+            if (template.pathMatch() == null && template.match() == null && template.getXContentFieldTypes().length == 0) {
+                throw new MapperParsingException("template [" + template.name() +
+                    " does not have [match] not [path_match] or [match_mapping_type] defined");
```
does not have [match] nor [path_match] nor [match_mapping_type] defined ?

dnhatn replied:
Good catch, fixed in https://github.com/elastic/elasticsearch/pull/69948/commits/8c9e4020ecf2bf37cd20039729ccacb73e457a9f

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:
```diff
@@ -400,18 +390,19 @@ private static void validateDynamicTemplate(Mapper.TypeParser.ParserContext pars
 
+        final XContentFieldType[] types = Arrays.stream(template.getXContentFieldTypes())
+            .filter(t -> template.isRuntimeMapping() == false ||t.supportsRuntimeField())
+            .toArray(XContentFieldType[]::new);
```
I was hoping that with one change I suggested above this will no longer be necessary, unless I am missing something.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:
```diff
@@ -400,18 +390,19 @@ private static void validateDynamicTemplate(Mapper.TypeParser.ParserContext pars
+
         Exception lastError = null;
-        boolean dynamicTemplateInvalid = true;
+        boolean dynamicTemplateInvalid = types.length > 0;
```
I am not following why this is needed, can you help me?

dnhatn replied:
I've simplified the existing validation code in https://github.com/elastic/elasticsearch/pull/69948/commits/00b92c15d6e66f8e6fe33953a03edf9107c31c3a.

javanna reviewed (commented):
I did another round of review and left some comments. This is real close: given how we reworked it I suggested possibly renaming and moving away from the "hint" terminology, because when a dynamic template name is specified in the bulk request, we expect it to be honored or the doc will not be indexed.

dnhatn committed:
fix message

dnhatn committed:
Remove "hint" from the parameter

dnhatn committed:
wording docs

dnhatn committed:
Add a test for dynamic template with a wrong type

dnhatn committed:
simplify match_mapping_type

dnhatn committed:
Simplify the validation

dnhatn committed:
minimize changes in serialization

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
more yaml tests

dnhatn commented:
@javanna I am happy with the current shape of the API. It's less confusing the previous iterations. Thank you so much for your feedback. I think I have addressed all your points.

javanna reviewed (commented):
I agree that this is ready, my last questions are around backwards compatibility and expected behaviour in a mixed cluster, LGTM besides that. @jpountz would you like to have a last quick look at the API and express your opinion on backwards compatibility?

dnhatn commented on server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:
```diff
@@ -251,12 +261,18 @@ public void parse(
                             throw new IllegalArgumentException("Update requests do not support versioning. " +
                                     "Please use `if_seq_no` and `if_primary_term` instead");
                         }
+                        // TODO: support dynamic_templates in update requests
```
I will work on it right after this PR.

dnhatn committed:
remove hint

dnhatn committed:
Reject dynamic_templates on old nodes

dnhatn commented on server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -655,6 +660,9 @@ private void writeBody(StreamOutput out) throws IOException {
         if (out.getVersion().onOrAfter(Version.V_7_10_0)) {
             out.writeBoolean(requireAlias);
         }
+        if (out.getVersion().onOrAfter(Version.V_8_0_0)) {
```
Yes, I think we should reject such requests. I added that BWC logic in https://github.com/elastic/elasticsearch/pull/69948/commits/b341fcc70c4baf5c631a1b74ac492878b30054e6.

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -711,10 +712,18 @@ private static void parseCopy(String field, ParseContext context) throws IOExcep
+                            "dynamic templates [" + context.sourceToParse().dynamicTemplates() + "]";
+                        throw new MapperParsingException("Field [" + currentPath + "] must be an object; " +
+                            "but it's configured as [" + fieldMapper.typeName() + "] in dynamic template [" +
+                            context.sourceToParse().dynamicTemplates().get(currentPath) + "]");
```
Sure, will do.

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
Add min node version

dnhatn committed:
stylecheck

dnhatn committed:
Revert "stylecheck"

This reverts commit 8a8084137c9d500f7321f976c40ff8a8a577ba62.

dnhatn committed:
Revert "Add min node version"

This reverts commit aef927e301f68b67ea3f44a3a1cbaee866377b17.

javanna commented:
I opened #70805 to discuss the mixed cluster version scenario more generally, and mention this PR as an example to see whether we are on the right track.

javanna commented:
We reached some conclusion on https://github.com/elastic/elasticsearch/issues/70805#issuecomment-805805047 . Do you have thoughts on this @dnhatn ?

dnhatn committed:
better error message

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:
```diff
@@ -400,18 +390,17 @@ private static void validateDynamicTemplate(Mapper.TypeParser.ParserContext pars
+                    parserContext.getIndexSettings().getIndex() + " created before [" + Version.V_8_0_0 + "]" +
+                    " must have [match] or [path_match] or [match_mapping_type] defined");
+            }
         }
```
In my opinion we should allow this on old indices the same way we allow adding new field types to old indices. Are we expecting something to break without this validation?

javanna replied:
I may missing something here but I was under the impression that dynamic template validation happens on every node as we parse the mappings. So, if we accepted the newer format we would end up failing when parsing the mappings on older nodes, and I am not terribly sure what effect that will have. It would be a good idea to test this, even manually, just to validate our assumptions.

javanna replied:
Also it seems consistent to accept a format when we know it is valid, rather than it causing errors down the line on some shards, again assuming that this is what happens in this case.

jpountz replied:
All nodes hold the cluster state, but we only parse mappings on nodes that hold a shard for the index that has these mappings. And the way that we avoid making everything fail in mixed-version cluster scenarios, is by requiring indices created by version N to be allocated on nodes whose version is >= N. Otherwise we would have the same problem whenever we are adding new field types or new options on existing field types. For new APIs or API parameters like `_pit` or the `dynamic_templates` parameter of bulk requests, we expect users of the cluster to wait until the cluster is fully upgraded before starting using them. This is why our upgrade instructions require upgrading Elasticsearch before Kibana and Beats/Agent.

javanna replied:
thanks for clarifying, I am aware of how shards are allocated based on their version, which is why it was suggested that we should reject the bulk request preemptively on older indices that don't support the new feature. I like how that approach guarantees consistency, meaning that when we accept the new parameter, we will apply it rather than failing on some shards (those allocated on the older nodes). This sounds like it is a good way to enforce that shippers don't use new features until all nodes are on the same upgraded version. What I was missing is that older indices that are allocated on newer nodes that support the feature will effectively not support the new feature, which I think is too restrictive. I am happy relying on the stream version here, under the assumption that mixed clusters are a temporary situation in which using a new feature is best effort.

javanna replied:
My comment mentioned the bulk request, while here we were specifically discussing mappings and adding the dynamic template in the newer format. Sorry about the confusion. You asked if we are expecting something to break without this validation: some nodes will fail parsing the mappings because of a format they don't understand, while others will be able to accept it, depending on where shards are allocated. I think what you are hinting at is that, despite this may sound inconsistent, it leads to a consistent behaviour, meaning that if at least one shard is allocated to an older version, the mapping update will be rejected. And that is acceptable in a rolling upgrade scenario, more than having to create a new index to use the new feature. I think I understand better now and I agree.

dnhatn replied:
Hi @javanna @jpountz 

Sorry for the late response. I was offline for the last week. I implemented a BWC check using the [min node version](https://github.com/elastic/elasticsearch/pull/69948/commits/aef927e301f68b67ea3f44a3a1cbaee866377b17). However, it is not correct because we [can't](https://github.com/elastic/elasticsearch/blob/a92a647b9f17d1bddf5c707490a19482c273eda3/server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java#L351) retrieve the cluster state from ClusterService while performing a cluster state update. I tried to pass the cluster state around, but the approach is invasive.

I looked at some implementations that introduce new mapping types. We assume that users won't use them until all nodes are upgraded; otherwise, some shards on old nodes will be failed and assigned to new nodes. I pushed https://github.com/elastic/elasticsearch/pull/69948/commits/60d11a4b2bdcd2cc6d1e86ad666449f008b4f140 to relax this validation so users can use this new format with old indices.

Are we okay to get this PR in as other new features? I will spend some time working on this validation later.

jpountz reviewed (commented):
I left some minor comments but other than that it looks good to me!

dnhatn committed:
Merge 'master' into dynamic-mapping-hint

dnhatn committed:
Merge branch 'master' into dynamic-mapping-hint

dnhatn committed:
move yaml test

dnhatn committed:
Update docs/reference/docs/bulk.asciidoc

Co-authored-by: Adrien Grand <jpountz@gmail.com>

dnhatn committed:
Allow new dynamic template format with old indices

javanna commented on docs/reference/docs/bulk.asciidoc:
```diff
@@ -738,3 +742,39 @@ The API returns the following result.
+      {
+        "geo_point": {
+         "mapping": {
+            "mapping" : "geo_point"
```
is this syntax correct? should it be type: geo_point ? Can you also fix the indentation?

dnhatn replied:
Thanks, fixed in https://github.com/elastic/elasticsearch/pull/69948/commits/d0010bd1e3546caebf82af6a3e446b23964fedcb

javanna commented on docs/reference/rest-api/common-parms.asciidoc:
```diff
@@ -631,6 +631,15 @@ If `true`, the destination must be an <<indices-aliases,index alias>>. Defaults
+Defaults to an empty map. If a name matches a dynamic template, then that template will be 
+applied regardless of other match predicates defined in the template. And if a field is 
+already defined in the mapping, then this parameter won't be used.
+end::bulk-dynamic-templates[]
```
this comment still needs addressing I think

dnhatn replied:
> Can you also make it explicit that the mappings will only be updated if the document succeeds indexing?

I am not sure if this always holds. An indexing operation can fail (for example: over doc count limit) after we perform a dynamic mapping update.

javanna commented on server/src/internalClusterTest/java/org/elasticsearch/index/mapper/DynamicMappingIT.java:
```diff
@@ -156,4 +171,119 @@ public void testMappingVersionAfterDynamicMappingUpdate() throws Exception {
+                    }
+                    if (randomBoolean()) {
+                        mappings.field("match_mapping_type", "string");
+                    }
```
I think I would try to avoid randomizing behaviour in an IT test. They tend to be hard to reproduce. If this is to have more coverage, shall we have one test method per scenario that we want to test?

dnhatn replied:
> They tend to be hard to reproduce.

We have a lot of improvements in IT tests recently, now they are almost reproducible as unit tests. Are you okay if I leave this test as is?

javanna replied:
ok

javanna reviewed (approved):
I left some minor comments, LGTM otherwise. Thanks for all the iterations.

dnhatn committed:
fix mapping syntax

dnhatn commented:
Thanks everyone :).
